### ISSUE TITLE
Remove deleted keys during memtable flushing for lock cf

### ISSUE THREAD

**Issue by youjiali1995** [tikv/tikv#8652]
## Description
The work pattern of lock CF in TiKV is insert one when prewrites and then remove it when commits and they are usually in the same memtable. However, RocksDB flushes all entries in the memtable to the disk which is wasteful for lock CF. You should implement it in https://github.com/tikv/rocksdb. Please be careful about the correctness of compaction in RocksDB.  

## SIG Slack Channel

You can join [#sig-transaction](https://tikv-wg.slack.com/archives/CU0TU3GSE) on slack in your spare time to discuss and get help with mentors or others.

## Score

* 5000

## Mentor(s)

* @youjiali1995  

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* C++
* RocksDB

## Learning Materials

* https://github.com/facebook/rocksdb/wiki/Compaction

**Comment by page4** [comment]
Why not use SingleDelete in RocksDB?

**Comment by youjiali1995** [comment]
> Why not use SingleDelete in RocksDB?

I didn't know that! I will take a look at it. Thanks!

**Comment by youjiali1995** [comment]
`SingleDelete` has such a constraint:
> SingleDelete() removes the database entry for a key, but has the prerequisites that the key exists and was not overwritten.

In TiKV, a key in lock CF may be overwritten multiple times, for example, increasing TTL, updating overlap field. If we want to use it, we need to `SingleDelete` it and then put it to overwrite a key in lock cf.

There is a compatible issue during the rolling upgrade. If a new version TiKV `SingleDelete`s a key and replicates the log to old version replicas, replicas can't apply it because they don't know how to handle it which results in data inconsistency.

Actually, I'm not sure whether it's worthy to do it and it's complicated. I think we need to figure out how much disk flow is consumed by the compaction of lock CF.

**Comment by youjiali1995** [comment]
Close it for now until we figure out how much disk flow is consumed by the compaction of lock CF.

**Comment by youjiali1995** [comment]
According to the comment https://github.com/tikv/tikv/issues/8140#issuecomment-862190703, the lock-cf compaction flow is high. 

![image](https://user-images.githubusercontent.com/14819777/160349347-99258fce-87a3-4713-a903-fdae3e27f4e5.png)

/cc @SpadeA-Tang

**Comment by Connor1996** [comment]
@youjiali1995 Seems what you want is `SingleDelete`? it's supported by RocksDB, and all CF can use it. But considering the reentability, we didn't use it.

**Comment by youjiali1995** [comment]
@Connor1996 Yes, we will make a naive POC to get its improvments before deciding whether to do it or not.

> But considering the reentability or idempotence, we didn't use it.

It indeed needs some work to solve these.

**Comment by Connor1996** [comment]
@youjiali1995 Okay, please let me know if you get the result
